### ISSUE TITLE
Move examples tests into their own test suite

### ISSUE THREAD

**Issue by lbernick** [tektoncd/plumbing#1045]
Examples tests are the flakiest test suite in Pipelines by far (see [Tekton Deflake](https://docs.google.com/document/d/1ur1-YF5uq85BRLveD2xjbWGyspxPUczseGQUPs1h8FI/edit#heading=h.rv8888vfg18n)). These should be moved into their own test suite, so we can separate out these flakes from other integration tests. In addition, they should probably not block PR submission.

**Comment by afrittoli** [comment]
Separate them out might reduce the overall execution time and reduce the "restest" time. 
I'd be hesitant in marking the job as non-voting though, since I think a lot of our e2e coverage is there.
I would rather move the flaky tests to the non-ci folder until they can be fixed.

**Comment by lbernick** [comment]
understandable! Let's start with moving them into their own test suite and getting data on the ratio of flakes to real failures-- then we can make a more informed decision. rewriting them to be more deterministic might also be helpful!

**Comment by lbernick** [comment]
I think we will first need a script in Pipelines responsible for running only examples tests (probably just a new function run_examples in the script you linked), then a new prow job (configured [here](https://github.com/tektoncd/plumbing/blob/main/prow/config.yaml) to run the examples tests. Finally, the examples tests can be moved out of the run_e2e function in the e2e test script. This issue is separate from the cluster that e2e tests run on (whether that is kind or gke).

**Comment by vdemeester** [comment]
> I think we will first need a script in Pipelines responsible for running only examples tests (probably just a new function run_examples in the script you linked), then a new prow job (configured [here](https://github.com/tektoncd/plumbing/blob/main/prow/config.yaml)) to run the examples tests. Finally, the examples tests can be moved out of the run_e2e function in the e2e test script. This issue is separate from the cluster that e2e tests run on (whether that is kind or gke).

I wonder if we need a new cron job. Saying that *mainly* because this would mean we use "another" cluster, if we don't use `kind`. I'd rather make sure we have tests yaml and examples test separate (like discussed in https://github.com/tektoncd/pipeline/issues/4810) and then figure out if we want to split that.

**Comment by afrittoli** [comment]
> I think we will first need a script in Pipelines responsible for running only examples tests (probably just a new function run_examples in the script you linked)

Sounds good - there is already an [env variable](https://github.com/tektoncd/pipeline/blob/7059cfd5e8a4f3619353aaf8f2bba722b70afa47/test/e2e-tests.sh#L26) about running the YAML tests or not, we could also add one about running the other tests.

> This issue is separate from the cluster that e2e tests run on (whether that is kind or gke).

It is, however adding a new prow job like the existing ones might be too expensive and might exhaust our boskos pool too quickly. Using `kind` would alleviate that as `kind` job do not use projects from the boskos pool.

The downside of using `kind` vs `gke` is that it's an extra change in the tests, which is not ideal when tracking down flaky tests. Perhaps we could convert the alpha flag test to be `kind` based, and spilt the regular test into to prow tests, so net balance would be the same amount of boskos accounts used.

**Comment by Yongxuanzhang** [comment]
Hi, as discussed with @lbernick, we can first move examples test into kind cluster, I think this could be part of this issue.(Like a first step.)
Here are some details I have in mind, correct me if I'm wrong since I'm quite new to the repo.

Similar like https://github.com/tektoncd/plumbing/pull/927 by @afrittoli, I can create several files:
- tekton/ci/
    - jobs
        - examples-kind.yaml (this will include a new script file to execute examples tests)
        - kustomization.yaml(add examples-kind.yaml)
    - pipeline (not sure if these can be merged into current files)
        - kustomization.yaml (add new files if needed)
        - examples-template (use examples params)
        - examples-trigger 
     - README: updated

 after this we can add new scripts in pipeline for examples test and disable it in e2e test

 Do we need to setup cronjob?

**Comment by Yongxuanzhang** [comment]
> > Hi, as discussed with @lbernick, we can first move examples test into kind cluster, I think this could be part of this issue.(Like a first step.) Here are some details I have in mind, correct me if I'm wrong since I'm quite new to the repo.
> > Similar like #927 by @afrittoli, I can create several files:
> > 
> > * tekton/ci/
> >   
> >   * jobs
> >     
> >     * examples-kind.yaml (this will include a new script file to execute examples tests)
> 
> I think you should be able to use the existing `e2e-kind.yaml` - I wrote that pipeline generic on purpose so that it may be used for any e2e job we want to run:
> 
> ```
>     - name: k8s-version
>       type: string
>       description: The version of k8s (e.g. v1.21.x, v1.22.x or v1.23.x)
>     - name: e2e-script
>       type: string
>       description: the path to the e2e script in the sources
>       default: "test/e2e-tests.sh"
>     - name: e2e-env
>       type: string
>       description: |
>         the path to an env file that is loaded before the execution
>         of the e2e-script
> ```
> 
> > ```
> > * kustomization.yaml(add examples-kind.yaml)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * pipeline (not sure if these can be merged into current files)
> >   
> >   * kustomization.yaml (add new files if needed)
> >   * examples-template (use examples params)
> 
> If you create separate templates, then you'll be forced to have dedicated triggers for them. Since the interceptors will be the same there is no need. Trigger template allow specifying a list of resources that needs to be created. In this case all you need is a new `PipelineRun` added to the list of resources in `template.yaml`, for the new job. It can use the same pipeline, but with different parameters and a different CI job
> 
> > ```
> > * examples-trigger
> > ```
> 
> I don't think we need new triggers, as all CI jobs shall be executed for the same incoming GitHub events.
> 
> > * README: updated
> > 
> > after this we can add new scripts in pipeline for examples test and disable it in e2e test
> > Do we need to setup cronjob?
> 
> If you want to run this job as a periodic job you will need a cronjob and a dedicated trigger for it. If you want to run it on every PR, you don't need anything extra.

Thanks a lot for the clarifications! I think I have more understanding of this! 😄 
Will draft a pr soon!

**Comment by lbernick** [comment]
/remove-lifecycle stale

Just chatted with @abayer -- all the pipelines e2e/integration tests now run on kind, so we are good to go with moving examples tests into their own test suite.

**Comment by abayer** [comment]
For what it's worth, I'd question whether this would be worth it - `pull-tekton-pipeline-integration-tests` takes ~22-25 minutes now, with the alpha job taking a couple minutes more, so there isn't much time to be saved by separating them out. I went that route initially with the kind jobs, but there's still a flat 3-4 minutes for spinning up the kind cluster and running `ko resolve` in each job, making the time savings not as pronounced as it could be. 

I'd still lean in that direction if it wasn't for the fact that we'd also be using 2 additional n2-standard-4 nodes for each PR push as well - the kind jobs are restricted to only run on n2-standard-4s (with expanded local storage) where each kind job uses the whole node. We've got a quota of 96 n2 CPUs currently, so that we can theoretically run 24 integration jobs at once, i.e., 12 PRs' worth. Splitting out examples would mean 4 jobs per PR push/etc, dropping us to 6 PRs' worth concurrently, and the fact that the combined runtime for the current two jobs would be a few minutes less than the combined runtime for the split-up four jobs means we would be spending a bit more as well.

So yeah, I tend to think we're ok with the current setup for now, but it's worth revisiting periodically.
